cmd/download-plugins/main.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	_ "embed"
+	"encoding/hex"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"aead.dev/minisign"
+	"github.com/google/go-github/v48/github"
+
+	"github.com/bufbuild/plugins/internal/release"
+)
+
+func main() {
+	if err := run(); err != nil {
+		log.Fatalf("failed to download release: %v", err)
+	}
+}
+
+func run() error {
+	var (
+		minisignPublicKey string
+		releaseTag        string
+	)
+	flag.StringVar(&minisignPublicKey, "minisign-public-key", "", "path to minisign public key file")
+	flag.StringVar(&releaseTag, "release-tag", "", "release to download (default: latest release)")
+	flag.Parse()
+
+	if len(flag.Args()) != 1 {
+		_, _ = os.Stderr.WriteString("usage: download-plugins <dir>\n")
+		flag.PrintDefaults()
+		os.Exit(2)
+	}
+	downloadDir := flag.Arg(0)
+
+	ctx := context.Background()
+	client := release.NewClient(ctx)
+
+	publicKey, err := loadPublicKey(minisignPublicKey)
+	if err != nil {
+		return err
+	}
+
+	var githubRelease *github.RepositoryRelease
+	if releaseTag == "" {
+		githubRelease, err = client.GetLatestRelease(ctx, release.GithubOwnerBufbuild, release.GithubRepoPlugins)
+		if err != nil {
+			return err
+		}
+	} else {
+		githubRelease, err = client.GetReleaseByTag(ctx, release.GithubOwnerBufbuild, release.GithubRepoPlugins, releaseTag)
+		if err != nil {
+			return err
+		}
+	}
+
+	pluginReleases, err := client.LoadPluginReleases(ctx, githubRelease, publicKey)
+	if err != nil {
+		return err
+	}
+	for _, pluginRelease := range pluginReleases.Releases {
+		exists, err := pluginExistsMatchingDigest(pluginRelease, downloadDir)
+		if err != nil {
+			return err
+		}
+		if exists {
+			log.Printf("already downloaded: %s", filepath.Join(downloadDir, filepath.Base(pluginRelease.URL)))
```

Comment:
If plugins are already downloaded to the download dir, we'll skip downloading them again if the SHA-256 digest matches:

```
2022/12/01 21:44:49 [DEBUG] GET https://api.github.com/repos/bufbuild/plugins/releases/latest
2022/12/01 21:44:49 [DEBUG] GET https://api.github.com/repos/bufbuild/plugins/releases/assets/86688755
2022/12/01 21:44:50 [DEBUG] GET https://api.github.com/repos/bufbuild/plugins/releases/assets/86688756
2022/12/01 21:44:50 already downloaded: download/apple-swift-v1.20.2.zip
2022/12/01 21:44:50 already downloaded: download/apple-swift-v1.20.3.zip
2022/12/01 21:44:50 already downloaded: download/bufbuild-connect-go-v1.0.0.zip
...
2022/12/01 21:44:51 already downloaded: download/protocolbuffers-ruby-v21.8.zip
2022/12/01 21:44:51 already downloaded: download/protocolbuffers-ruby-v21.9.zip
2022/12/01 21:44:51 already downloaded: download/protocolbuffers-ruby-v21.10.zip
```